File: 04-polimorfismo/python/circulo.py
"""
circulo.py - Figura + Dibujable

La herencia multiple de Python hace en una linea lo que Java separa
en "extends" (una sola clase) e "implements" (muchas interfaces).
"""

import math

from dibujable import Dibujable
from figura import Figura


class Circulo(Figura, Dibujable):

    def __init__(self, radio):
        super().__init__("Circulo")
        self._radio = radio

    def area(self):
        return math.pi * self._radio ** 2

    def perimetro(self):
        return 2 * math.pi * self._radio

    def dibujar(self):
        print(f"Dibujando un circulo de radio {self._radio}  ( O )")

File: 04-polimorfismo/java/Circulo.java
/*
 * Circulo.java
 *
 * "extends Figura implements Dibujable" se lee:
 *    ES UNA Figura   (herencia: hereda nombre y getNombre)
 *    y ademas FIRMA el contrato Dibujable
 */
public class Circulo extends Figura implements Dibujable {

    private double radio;

    public Circulo(double radio) {
        super("Circulo");
        this.radio = radio;
    }

    @Override
    public double area() {
        return Math.PI * radio * radio;
    }

    @Override
    public double perimetro() {
        return 2 * Math.PI * radio;
    }

    @Override
    public void dibujar() {
        System.out.println("Dibujando un circulo de radio " + radio + "  ( O )");
    }
}
